src/main/scala/org/broadinstitute/dsde/workbench/sam/schema/JndiSchemaDAO.scala

```diff
@@ -49,40 +52,190 @@ object JndiSchemaDAO {
   }
 }
 
-class JndiSchemaDAO(protected val directoryConfig: DirectoryConfig)(implicit executionContext: ExecutionContext) extends JndiSupport with DirectorySubjectNameSupport {
+object SchemaStatus {
+  sealed trait SchemaStatus
+
+  case object Wait extends SchemaStatus
+  case object Proceed extends SchemaStatus
+  case object Ignore extends SchemaStatus
+}
+
+class JndiSchemaDAO(protected val directoryConfig: DirectoryConfig, val schemaLockConfig: SchemaLockConfig)(implicit executionContext: ExecutionContext) extends JndiSupport with DirectorySubjectNameSupport with LazyLogging {
 
   def init(): Future[Unit] = {
+    if(schemaLockConfig.lockSchemaOnBoot) { initWithSchemaLock() }
+    else { createAndDestroySchema().map(_ => ()) }
+  }
+
+  private def initWithSchemaLock(): Future[Unit] = {
+    for {
+      schemaLockStatus <- initSchemaLock()
+      schemaUpdateStatus <- schemaLockStatus match {
```

Comment:
what if this step has a failure? Should we remove the lock record?

Reply:
My initial thought was no. For comparison, Liquibase _does_ remove the lock when something fails, but schema updates are incremental and there are a lot more safety checks in place.

In our case, I think if destroyAndCreateSchema fails, it requires a bit more examination because many things could have gone wrong and there are fewer protections in place. I think forcing some additional human intervention is better but maybe that's just flat-out annoying.

I could be easily swayed to the other side of this argument.